### PR TITLE
Adding codegen updates and configuration options

### PR DESCRIPTION
**Pull request by davidh44**
## Motivation and Context

Adding codegen updates and configuration options for the RequestCompression trait

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [ ] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Comment by L-Applin** [comment]
I don't see any unit tests for the new codegen, is it not a good time to add some?

**Comment by davidh44** [comment]
> I don't see any unit tests for the new codegen, is it not a good time to add some?

Yeah planning to add tests at the end after implementation

**Review by L-Applin — APPROVED** [review]

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestCompressionConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+     * If set, returns true if request compression is enabled, else false if request compression is disabled.
+     */
+    public Boolean requestCompressionEnabled() {
+        return requestCompressionEnabled;
+    }
+
+    /**
+     * If set, returns the minimum compression threshold in bytes, inclusive, in order to trigger request compression.
+     */
+    public Integer getMinimumCompressionThresholdInBytes() {
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestCompressionConfiguration.java:49]: fluent style

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestCompressionConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        return result;
+    }
+
+
+    public interface Builder extends CopyableBuilder<Builder, RequestCompressionConfiguration> {
+
+        /**
+         * Configures whether request compression is enabled or not.
+         *
+         * @param requestCompressionEnabled
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestCompressionConfiguration.java:95]: Can we add a note about the default behavior? Same for the other setting

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestOverrideConfiguration.java [hunk 3]
```diff
@@ -412,6 +425,15 @@ default B putRawQueryParameter(String name, String value) {
 
         ExecutionAttributes executionAttributes();
 
+        /**
+         * Sets the {@link RequestCompressionConfiguration} for this request.
+         *
+         * @param requestCompressionConfiguration Request compression configuration object for this request.
+         */
+        B requestCompressionConfiguration(RequestCompressionConfiguration requestCompressionConfiguration);
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestOverrideConfiguration.java:433]: Consumer builder pattern overload?

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/SdkSystemSetting.java [hunk 4]
```diff
@@ -184,6 +184,10 @@ public enum SdkSystemSetting implements SystemSetting {
      */
     AWS_USE_FIPS_ENDPOINT("aws.useFipsEndpoint", null),
 
+    AWS_REQUEST_COMPRESSION_ENABLED("aws.requestCompressionEnabled", null),
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/SdkSystemSetting.java:187]: Let's add some javadoc

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.java [hunk 5]
```diff
@@ -230,19 +234,30 @@ public ExecutionAttributes executionAttributes() {
...
+                       .add("retryPolicy", retryPolicy)
+                       .add("apiCallTimeout", apiCallTimeout)
+                       .add("apiCallAttemptTimeout", apiCallAttemptTimeout)
+                       .add("executionInterceptors", executionInterceptors)
+                       .add("advancedOptions", advancedOptions)
+                       .add("profileFile", defaultProfileFile)
+                       .add("profileName", defaultProfileName)
+                       .add("scheduledExecutorService", scheduledExecutorService)
+                       .add("requestCompressionConfiguration", requestCompressionConfiguration)
+                       .build();
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.java:260]: Did we update hashcode and equals?

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.java [hunk 6]
```diff
@@ -230,19 +234,30 @@ public ExecutionAttributes executionAttributes() {
         return executionAttributes;
     }
 
+    /**
+     * The request compression configuration object, which includes options to enable/disable request compression and set the
+     * minimum compression threshold.
+     *
+     * @see Builder#requestCompressionConfiguration(RequestCompressionConfiguration)
+     */
+    public RequestCompressionConfiguration requestCompressionConfiguration() {
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.java:243]: I think this class returns optional for unset properties, we should probably be consistent.

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/trait/RequestCompression.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    private RequestCompression(Builder builder) {
+        this.encodings = builder.encodings;
+        this.isStreaming = builder.isStreaming;
+    }
+
+    public List<String> getEncodings() {
+        return encodings;
+    }
+
+    public void setEncodings(List<String> encodings) {
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/trait/RequestCompression.java:38]: Ideally, all classes should be immutable (except codegen module)

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.java [hunk 8]
```diff
@@ -80,6 +82,7 @@ private ClientOverrideConfiguration(Builder builder) {
         this.metricPublishers = Collections.unmodifiableList(new ArrayList<>(builder.metricPublishers()));
         this.executionAttributes = ExecutionAttributes.unmodifiableExecutionAttributes(builder.executionAttributes());
         this.scheduledExecutorService = builder.scheduledExecutorService();
+        this.requestCompressionConfiguration = builder.requestCompressionConfiguration();
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.java:85]: Did we add profile config option for `requestCompressionConfiguration`? I don't see it in this PR. Side note: it'd be great if we could create separate PRs for different features to make reviews eaiser, for example, separate PRs for codegen update and config options.

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestOverrideConfiguration.java [hunk 9]
```diff
@@ -153,6 +155,15 @@ public ExecutionAttributes executionAttributes() {
         return executionAttributes;
     }
 
+    /**
+     * Returns the request compression configuration object, which includes options to enable/disable request compression and
+     * set the minimum compression threshold. This request compression config object supersedes the request compression config
+     * object set on the client.
+     */
+    public RequestCompressionConfiguration requestCompressionConfiguration() {
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestOverrideConfiguration.java:163]: Same here, i think this class returns Optional for unset properties

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestOverrideConfiguration.java [hunk 10]
```diff
@@ -412,6 +425,26 @@ default B putRawQueryParameter(String name, String value) {
 
         ExecutionAttributes executionAttributes();
 
+        /**
+         * Sets the {@link RequestCompressionConfiguration} for this request.
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestOverrideConfiguration.java:429]: Can we mention precedence here since this method may be what users check first

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.java (hunk 5 above)
Comment (zoewangg) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.java:260]: Hmm, I guess that's an oversight. I guess we can leave it as is for now, not sure if adding them now will break things...

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/SdkExecutionAttribute.java [hunk 11]
```diff
@@ -109,6 +110,8 @@ public class SdkExecutionAttribute {
     public static final ExecutionAttribute<ChecksumValidation> HTTP_RESPONSE_CHECKSUM_VALIDATION = new ExecutionAttribute<>(
         "HttpResponseChecksumValidation");
 
+    public static final ExecutionAttribute<RequestCompressionConfiguration> REQUEST_COMPRESSION_CONFIGURATION =
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/SdkExecutionAttribute.java:113]: Missing Javadoc

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/SdkInternalExecutionAttribute.java [hunk 12]
```diff
@@ -92,6 +93,12 @@ public final class SdkInternalExecutionAttribute extends SdkExecutionAttribute {
     public static final ExecutionAttribute<Boolean> IS_DISCOVERED_ENDPOINT =
         new ExecutionAttribute<>("IsDiscoveredEndpoint");
 
+    /**
+     * The supported compression algorithms for an operation
+     */
+    public static final ExecutionAttribute<RequestCompression> REQUEST_COMPRESSION =
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/SdkInternalExecutionAttribute.java:99]: Why do we need it here if it's already in `SdkExecutionAttribute`?

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/traits/RequestCompressionTrait.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+    private RequestCompressionTrait() {
+    }
+
+    /**
+     * Generate a ".putExecutionAttribute(...)" code-block for the provided operation model. This should be used within the
+     * context of initializing {@link ClientExecutionParams}. If request compression is not required by the operation, this will
+     * return an empty code-block.
+     */
+    public static CodeBlock create(OperationModel operationModel) {
```
Comment (zoewangg) [inline codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/traits/RequestCompressionTrait.java:41]: Let's add some tests to show generated code.

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestCompressionConfiguration.java [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+        }
+
+        @Override
+        public Builder requestCompressionEnabled(Boolean requestCompressionEnabled) {
+            this.requestCompressionEnabled = requestCompressionEnabled;
+            return this;
+        }
+
+        @Override
+        public Builder minimumCompressionThresholdInBytes(Integer minimumCompressionThresholdInBytes) {
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestCompressionConfiguration.java:129]: Should this be Long to be future proof?

File: core/profiles/src/main/java/software/amazon/awssdk/profiles/ProfileProperty.java [hunk 15]
```diff
@@ -141,6 +141,17 @@ public final class ProfileProperty {
 
     public static final String EC2_METADATA_SERVICE_ENDPOINT = "ec2_metadata_service_endpoint";
 
+    /**
+     * Whether request compression is enabled for operations marked with the RequestCompression trait. The default value is true.
+     */
+    public static final String REQUEST_COMPRESSION_ENABLED = "requestCompressionEnabled";
```
[the remark this answers is not in the document]
Comment (davidh44) [reply core/profiles/src/main/java/software/amazon/awssdk/profiles/ProfileProperty.java]: updated values for profile and env var to be compliant with specification

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestCompressionConfiguration.java (hunk 14 above)
Comment (davidh44) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestCompressionConfiguration.java:129]: discussed offline, specification states integer value as must

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/query/service-2.json [hunk 16]
```diff
@@ -59,6 +59,17 @@
...
+      "name": "APostOperation",
+      "http": {
+        "method": "POST",
+        "requestUri": "/"
+      },
+      "authtype": "none",
+      "requestCompression": {
+        "encodings": ["gzip"]
+      }
+    },
```
Comment (zoewangg) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/query/service-2.json:72]: Thanks for adding the test. I think we should add tests for other protocols as well since it seems the codegen code for different protocols is in different files.

**Review by zoewangg — APPROVED** [review]
